### PR TITLE
Add docs describing saved tensor hooks

### PR DESCRIPTION
**Pull request by Varal7**
Add section to the Autograd mechanics docs to describe the recently
exposed saved tensors (#52451), how to register packing / unpacking
hooks (#60975) and how to use default hooks (#61834)

Sister PR: #62361 (will add a link from autograd.rst to notes/autograd in whatever PR does not land first)

### CODE REVIEW

File: docs/source/notes/autograd.rst [hunk 1]
```diff
@@ -36,6 +36,51 @@ flow statements, that can change the overall shape and size of the graph at
...
+
+Some operations need intermediary results to be saved during the forward pass
+in order to execute the backward pass. For example, the function
+:math:`x\mapsto x^2` needs to save the input :math:`x` because it is needed to
+compute the gradient.
+
+When defining a custom Python :class:`~torch.autograd.Function`, you can use
+:func:`~torch.autograd.function._ContextMethodMixin.save_for_backward` to save
+tensors during the forward pass and
+:attr:`~torch.autograd.function.NestedIOFunction.saved_tensors` to retrieve them
```
Comment (albanD) [inline docs/source/notes/autograd.rst:52]: That shouldn't point to the `NestedIOFunction` version here rigth?

File: docs/source/notes/autograd.rst [hunk 2]
```diff
@@ -36,6 +36,51 @@ flow statements, that can change the overall shape and size of the graph at
...
+
+.. code::
+
+    x = torch.randn(5, requires_grad=True)
+    y = x.pow(2)
+    print(x.equal(y.grad_fn._saved_self)) # True
+    print(x is y.grad_fn._saved_self) # True
+
+
+In the previous code, ``y.grad_fn._saved_self`` shares the same storage as `x`.
```
Comment (albanD) [inline docs/source/notes/autograd.rst:68]: Actually they do share the same storage. And they always do, except for inplace ops that save their input (that is being modified inplace and so obviously we need another Tensor with different storage to store the original value).
Here you want to say that they might not be the same Tensor object.

File: docs/source/notes/autograd.rst [hunk 3]
```diff
@@ -598,3 +643,83 @@ chain rule:
...
+
+.. _saved-tensors-hooks-doc:
+
+Hooks for saved tensors
+-----------------------
+
+You can control :ref:`how saved tensors are packed / unpacked
+<saved-tensors-doc>` by defining a pair of ``pack_hook`` / ``unpack_hook`` hooks.
+The ``pack_hook`` function should take a tensor as its single argument but can
+return any python object (e.g. another tensor, a tuple, or even a filename).
```
Comment (albanD) [inline docs/source/notes/autograd.rst:655]: nit: the third element in your or is not a type. Maybe "or even a string containing a filename"?

File: docs/source/notes/autograd.rst [hunk 4]
```diff
@@ -598,3 +643,83 @@ chain rule:
...
+The ``unpack_hook`` function takes as its single argument the output of
+``pack_hook`` and should return the original tensor to be used in the backward
```
Comment (albanD) [inline docs/source/notes/autograd.rst:657]: "the original tensor" -> we should define that better here.
It can actually be a different Tensor object. It can have a different storage, etc

The requirement is that it should be a Tensor that has the same content.
Any autograd-related metadata should be ignored in particular as they will be overwritten.

File: docs/source/notes/autograd.rst [hunk 5]
```diff
@@ -598,3 +643,83 @@ chain rule:
...
+        torch.save(x, name)
+        return name
+
+    def unpack_hook(name):
+        return torch.load(name)
+
+.. warning::
+
+    Both functions should avoid performing inplace operations on the input or
+    undefined behavior may arise.
```
Comment (albanD) [inline docs/source/notes/autograd.rst:675]: I am not a fan of have a UB here.
Why is inplace during packing forbidden?

Inplace during unpacking, we should say that this is forbidden and this function should be callable multiple times and should return the same data everytime.

File: docs/source/notes/autograd.rst [hunk 6]
```diff
@@ -598,3 +643,83 @@ chain rule:
...
+The ``unpack_hook`` function takes as its single argument the output of
+``pack_hook`` and should return the original tensor to be used in the backward
+pass.
+
+An example of such pair is:
+
+.. code::
+
+    def pack_hook(x):
+        name = os.path.join(tmp_dir, str(uuid.uuid4()))
```
Comment (albanD) [inline docs/source/notes/autograd.rst:665]: I like this example, but we should use it to show how to handle the lifetime as well I think.
In particular, with this code, you're gonna "leak" everything you write on disk.

I think you want to show that the file on disk should live as long as the object your returned. And that the unpack must NOT delete the file as it might be called multiple times.

File: docs/source/notes/autograd.rst [hunk 7]
```diff
@@ -598,3 +643,83 @@ chain rule:
...
+The ``pack_hook`` method is called as soon as the pair is registered.
+The ``unpack_hook`` method is called each time the saved tensor needs to be
+accessed, either by means of ``y.grad_fn._saved_self`` or during the backward
+pass.
+
+.. warning::
+
+    If you maintain a reference to a :class:`SavedTensor` after the saved
+    tensors have been released (i.e. after backward has been called) and call
+    its :meth:`~torch.autograd.SavedTensor.register_hooks`, undefined behavior
```
Comment (albanD) [inline docs/source/notes/autograd.rst:701]: nit: mention explicitly that this is forbidden I think. And then say that if the user still does it, it will be UB as we cannot catch that call all the time.

File: docs/source/notes/autograd.rst [hunk 8]
```diff
@@ -598,3 +643,83 @@ chain rule:
...
+    may arise. Pytorch will throw an error most of the time but it may fail
+    silently in some cases.
+
+Registering default hooks for saved tensors
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Alternatively, you can register a pair of default hooks which will be applied
+to all saved tensors that are created in that context.
```
Comment (albanD) [inline docs/source/notes/autograd.rst:709]: Do you want to give an example of this here?

Also maybe show how the hook can filter so that it only moves to disk Tensors that have more than "threshold" elements in them.

File: docs/source/notes/autograd.rst [hunk 9]
```diff
@@ -598,3 +643,83 @@ chain rule:
...
+
+Note that using a pair of identity hooks is not a no-op as it forces
+packing / unpacking even in cases where Pytorch would have chosen to not
```
Comment (albanD) [inline docs/source/notes/autograd.rst:712]: nit: I would phrase this slightly differently. More like having a pair of hooks disables all the optimization in place to reduce Tensor object creation.

**Review by albanD — COMMENTED** [review]
Looks quite good!
Just some small comments.

File: docs/source/notes/autograd.rst (hunk 5 above)
Comment (Varal7) [reply docs/source/notes/autograd.rst:675]: For example:

```python
def pack(t):
    t += 2
    return t - 2

torch.autograd.graph.set_saved_tensors_default_hooks(pack, lambda x: x)
a = torch.ones(5, requires_grad=True)
print(a)  # tensor([1., 1., 1., 1., 1.], requires_grad=True)
y = a * a
print(a)  # tensor([5., 5., 5., 5., 5.], requires_grad=True)
```

I guess that's a case where we can actually catch the inplace modification. The question is whether we prefer making inplace operations during pack forbidden or UB.

File: docs/source/notes/autograd.rst (hunk 6 above)
Comment (Varal7) [reply docs/source/notes/autograd.rst:665]: The solution I have in mind is to clear the tmp_dir after each backward. Is there a better way you were thinking?

File: docs/source/notes/autograd.rst (hunk 5 above)
Comment (albanD) [reply docs/source/notes/autograd.rst:685]: It's not UB here. It actually does modify the input inplace. But it doesn't lead to any silently wrong gradients beyond that.
But I do agree that the user shouldn't do that  and we want to forbid it :D

File: docs/source/notes/autograd.rst (hunk 6 above)
Comment (albanD) [reply docs/source/notes/autograd.rst:669]: You can have a file handle here that will delete the file when the handle is deleted. You can check the python's TempFile, I think they do that.

File: docs/source/notes/autograd.rst (hunk 5 above)
Comment (Varal7) [reply docs/source/notes/autograd.rst:685]: I'd argue that in the correct form, it's an UB because the way the backward formula was implemented can be subject to change without us notifying the user, as it is an implementation detail that they should not rely on. 
I can make a PR to always throw an error when the tensor is modified and make it very well defined behavior :)

File: docs/source/notes/autograd.rst [hunk 10]
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
...
 
+.. _saved-tensors-doc:
+
+Saved tensors
+^^^^^^^^^^^^^
+
+Some operations need intermediary results to be saved during the forward pass
+in order to execute the backward pass. For example, the function
+:math:`x\mapsto x^2` needs to save the input :math:`x` because it is needed to
+compute the gradient.
```
Comment (zou3519) [inline docs/source/notes/autograd.rst:47]: Nit: it could technically save `x^2` so I wouldn't say it "needs to save the input". Maybe just "the function saves the input x to compute the gradient"

File: docs/source/notes/autograd.rst [hunk 11]
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
...
+
+When defining a custom Python :class:`~torch.autograd.Function`, you can use
+:func:`~torch.autograd.function._ContextMethodMixin.save_for_backward` to save
+tensors during the forward pass and
+:attr:`~torch.autograd.function.Function.saved_tensors` to retrieve them
+during the backward pass. See :doc:`/notes/extending` for more information.
+
+For operations that Pytorch defines (e.g. ``aten::pow``), tensors are
```
Comment (zou3519) [inline docs/source/notes/autograd.rst:55]: nit: Pytorch -> PyTorch.

Also should we talk about `torch.pow` instead of `aten::pow`?

File: docs/source/notes/autograd.rst [hunk 12]
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
...
+automatically saved as they are needed. You can explore which tensors are saved
+by a certain ``grad_fn`` by looking for its attributes starting with the prefix
+``_saved``.
+
+.. code::
+
+    x = torch.randn(5, requires_grad=True)
+    y = x.pow(2)
+    print(x.equal(y.grad_fn._saved_self)) # True
+    print(x is y.grad_fn._saved_self) # True
```
Comment (zou3519) [inline docs/source/notes/autograd.rst:65]: Nit: Python style guide says two spaces have to exist before the start of an inline comment

File: docs/source/notes/autograd.rst [hunk 13]
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
...
+When defining a custom Python :class:`~torch.autograd.Function`, you can use
+:func:`~torch.autograd.function._ContextMethodMixin.save_for_backward` to save
+tensors during the forward pass and
+:attr:`~torch.autograd.function.Function.saved_tensors` to retrieve them
+during the backward pass. See :doc:`/notes/extending` for more information.
+
+For operations that Pytorch defines (e.g. ``aten::pow``), tensors are
+automatically saved as they are needed. You can explore which tensors are saved
+by a certain ``grad_fn`` by looking for its attributes starting with the prefix
+``_saved``.
```
Comment (zou3519) [inline docs/source/notes/autograd.rst:58]: I assume we prefixed this with `_` because we don't want users to rely on it. Is this true? If so we should explain that these are for debugging only

File: docs/source/notes/autograd.rst (hunk 5 above)
Comment (Varal7) [reply docs/source/notes/autograd.rst:699]: Here's a concrete example, powered by the fact that different backends have different derivative formulas for a given op.

```python
def f(device):
  lstm = nn.LSTM(5, hidden_size=5, num_layers=2, bidirectional=True, batch_first=True).to(device)
  for x in lstm.parameters():
    x.data.zero_()
  x = torch.zeros(5, 5, 5).to(device)
  return lstm(x)[0].sum()

print(f("cpu"))   # tensor(0., grad_fn=<SumBackward0>)
print(f("cuda"))  # tensor(0., device='cuda:0', grad_fn=<SumBackward0>)

def inc(x):
  x += 1
  return x - 1

torch.autograd.graph.set_saved_tensors_default_hooks(inc, lambda x: x)
print(f("cpu"))   # tensor(2700., grad_fn=<SumBackward0>)
print(f("cuda"))  # tensor(497.9154, device='cuda:0', grad_fn=<SumBackward0>)
torch.autograd.graph.reset_saved_tensors_default_hooks()
```

Collab: https://colab.research.google.com/drive/1XD9D2ack7WIAyrUYujuW0oKm8VYLkqAF?usp=sharing

File: docs/source/notes/autograd.rst [hunk 14]
```diff
@@ -36,6 +36,57 @@ flow statements, that can change the overall shape and size of the graph at
...
+
+Saved tensors
+^^^^^^^^^^^^^
+
+Some operations need intermediary results to be saved during the forward pass
+in order to execute the backward pass. For example, the function
+:math:`x\mapsto x^2` saves the input :math:`x` to compute the gradient.
+
+When defining a custom Python :class:`~torch.autograd.Function`, you can use
+:func:`~torch.autograd.function._ContextMethodMixin.save_for_backward` to save
```
Comment (albanD) [inline docs/source/notes/autograd.rst:49]: cc @soulitzer this link will need updating when we rename that class

**Review by albanD — APPROVED** [review]
LGTM!
Thanks for the update following the latest PRs
